src/navigation/reroute.js

```diff
@@ -92,6 +92,15 @@ export function reroute(pendingPromises = [], eventArguments) {
 
       const unmountAllPromise = Promise.all(allUnmountPromises);
 
+      unmountAllPromise.then(() => {
```

Comment:
We have several things calling `unmountAllPromise.then()` (see code below), but this is the first one. Which I think guarantees that the single-spa:before-mount-routing-event will be called before any mounting begins. Not 100% on that though.

The other thing, that I'm even less confident about, is where browser events end up in the event loop. We're calling dispatchEvent right before we start mounting the apps in a microtask. But I'm not sure whether the event handler gets called before or after those microtasks begin to mount the apps.

Reply:
That is encouraging. I think this makes sense to add. I agree that moving the DOM nodes between unmounting and mounting makes the most sense for perf. I wonder if it would make sense to make a more drastic change for additional guarantees

`unmountPromise.then(movePromise).then(mountPromise)`

Where movePromise is a no-op unless it's provided to `single-spa` by `single-spa-layout`

Reply:
Interesting idea.

> I agree that moving the DOM nodes between unmounting and mounting makes the most sense for perf. 

It's not for performance, but for correctness. If we rearrange the dom elements at the wrong moment, it will look janky to the user because applications will jump around the page.

> I wonder if it would make sense to make a more drastic change for additional guarantees

The event loop is predictable, and [this codepen](https://codepen.io/joeldenning/pen/LYpjwmR?editors=0010) shows that we already have the guarantees. I checked it in FF, Chrome, and Safari, and the dispatched event is handled BEFORE the next microtask. Dispatching events occurs synchronously, so we have guaranteed timing of when single-spa-layout will rearrange the dom elements.

Reply:
> It's not for performance, but for correctness. If we rearrange the dom elements at the wrong moment, it will look janky to the user because applications will jump around the page.

🤯  I didn't even consider that. I was thinking about moving huge trees of nodes and the perf implications that would have on the DOM but what you mentioned is more important.

> The event loop is predictable, and this codepen shows that we already have the guarantees when using dispatchEvent. I checked it in FF, Chrome, and Safari, and the dispatched event is handled BEFORE the next microtask. 

Awesome. I have no qualms about adding this new event to single-spa. I think it will prove extremely useful for `single-spa-layout` and perhaps more things down the road.